src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,3653 @@
+  println(io, "Array of length ", length(F))
+  for i=1:length(F)
+    if isassigned(F.O, i)
+      println(io, i, " -> ", F.O[i])
+    end
+  end
+  if isdefined(F, :S)
+    println(io, "defined on the Singular side")
+  end
+end
+
+@doc Markdown.doc"""
+    length(F::ModuleGens)
+
+Return the number of elements of the module generating set.
+Note: This is not the length in the mathematical sense!
+"""
+length(F::ModuleGens) = length(oscar_generators(F))
+
+@doc Markdown.doc"""
+    ngens(F::ModuleGens)
+
+Return the number of elements of the module generating set.
+"""
+ngens(F::ModuleGens) = length(oscar_generators(F))
+
+# i-th entry of module generating set on Oscar side
+# Todo: clean up, convert or assure
+function getindex(F::ModuleGens, ::Val{:O}, i::Int)
+  if !isassigned(F.O, i)
+    F.O[i] = convert(F.F, singular_generators(F)[i])
+  end
+  return oscar_generators(F)[i]
+end
+
+# i-th entry of module generating set on Singular side
+# Todo: clean up, convert or assure
+function getindex(F::ModuleGens, ::Val{:S}, i::Int)
+  if !isdefined(F, :S)
+    F.S = Singular.Module(base_ring(F.SF), [convert(F.SF,x) for x = oscar_generators(F)]...)
+  end
+  return F.S[i]
+end
+
+@doc Markdown.doc"""
+    oscar_assure(F::ModuleGens)
+
+If fields of `F` from the OSCAR side are not defined, they
+are computed, given the Singular side.
+"""
+function oscar_assure(F::ModuleGens)
+  if !isdefined(F, :O)
+    F.O = [convert(F.F, singular_generators(F)[i]) for i=1:Singular.ngens(singular_generators(F))]
+  end
+end
+
+@doc Markdown.doc"""
+    singular_assure(F::ModuleGens)
+
+If fields of `F` from the Singular side are not defined, they
+are computed, given the OSCAR side.
+"""
+function singular_assure(F::ModuleGens)
+  if !isdefined(F, :S)
+    if length(F) == 0
+      singular_ring = base_ring(F.SF)
+      F.S = Singular.Module(singular_ring, Singular.vector(singular_ring, singular_ring(0)))
+      return 
+    end
+    F.S = Singular.Module(base_ring(F.SF), [convert(F.SF,x) for x = oscar_generators(F)]...)
+    return
+  end
+  #F[Val(:S), 1]
+end
+
+# i-th entry of module generating set (taken from Oscar side)
+getindex(F::ModuleGens, i::Int) = getindex(F, Val(:O), i)
+
+@doc Markdown.doc"""
+    singular_module(F::FreeMod)
+
+Create a Singular module from an OSCAR free module.
+"""
+function singular_module(F::FreeMod)
+  Sx = singular_ring(base_ring(F), keep_ordering=false)
+  return Singular.FreeModule(Sx, dim(F))
+end
+
+@doc Markdown.doc"""
+    convert(SF::Singular.FreeMod, m::FreeModElem)
+
+Convert an OSCAR free module element to the Singular side.
+"""
+function convert(SF::Singular.FreeMod, m::FreeModElem)
+  g = Singular.gens(SF)
+  e = SF()
+  Sx = base_ring(SF)
+  for (p,v) = m.coords
+    e += Sx(v)*g[p]
+  end
+  return e
+end
+
+@doc Markdown.doc"""
+    convert(F::FreeMod, s::Singular.svector)
+
+Convert a Singular vector to a free module element on the OSCAR side.
+"""
+function convert(F::FreeMod, s::Singular.svector)
```

Comment:
Hmmm, I think this might be problematic (this also goes for the matching function in `FreeModules-graded.jl`, i.e., @fieker should perhaps adapt that, too):  `convert(F, s)` where `F` is an object (not a type) is not what Julia normally does, and I think we mostly replaced this by `F(s)` elsewhere?